### DISCUSSION TITLE
Alternative hysterisis for fan off point?

### DISCUSSION THREAD

**Discussion by ionGL** [Rem0o/FanControl.Releases#2320] [Ideas]
I was thinking recently about this.

Check out this graph as an example. The fan is off until 55.9C and at 56C it starts spinning at 50%. 

![gpugraph](https://github.com/Rem0o/FanControl.Releases/assets/80151979/5933a710-67c5-4c44-9a8e-83de19392ed1)

Would a setting with this logic make sense? "Hey, once the fan went from the zero % point to an on point, on the way down, don't activate the zero point if the temp hasn't dropped whatever user selectable degrees below the first activation point or below a selectable temp".

I want the fan to be reasonably responsive up and down when the load temperature fluctuates, but activate the fan off a bit later.

We see this with nvidia cards. The fan kicks in at 48C and goes up as needed, then as temps drop, it returns to the starting fan % to cool it down to 40C before killing the fan.

I can't have a graph with lower % points, as 0-30% are still 0%, and the fans need 50% to start and stay on reliably.

**Comment by Rem0o** [comment]
What you are describing most ressemble a trigger.

See https://getfancontrol.com/docs/

**Reply by ionGL** [reply]
Damn, i've been away for a year and forgot about this!

It makes perfect sense. You max mix a graph with a trigger and you get exactly the behavior described. Hysteresis in the graph isn't helping much when you want to control the gpu fan, as the gpu temp could drop 15-20c the moment you exit a game. So if you were to put a high hysteresis value, you are stuck with the highest % latched, even if the temp drops a few degrees as you game.

This way, trigger takes over when the gpu temp drops bellow the graph fan start/stop point and runs the fan at the desired % to cool it further.

![trigger](https://github.com/Rem0o/FanControl.Releases/assets/80151979/9f4446b9-c66d-4329-96aa-d8efca5870bd)

**Reply by Rem0o** [reply]
Yup you got it.

**Reply by Andss84** [reply]
I am interested as well well but I didn't really get what you did. why you have the sensor with a 10s average and why you use it for the Gpu graph as well besides the trigger?

**Reply by ionGL** [reply]
The 10s average is a way to delay a bit the fan reaction time. You can ignore that and use the real time gpu temp value.

![fancontrol](https://github.com/Rem0o/FanControl.Releases/assets/80151979/59267bbd-4b12-40af-910b-459f4eca91ca)

Here is my current setup for the gpu. You should read the Trigger curve explanation - https://getfancontrol.com/docs/

> The trigger fan curve will hold its fan speed until one of the two changing point is triggered. Below the idle temperature, idle fan speed is applied until the temperature gets back up to the load temperature. Above the load temperature, load fan speed is applied until the temperature gets back down to the idle temperature.

> 


For the gpu fans we use the mix curve, with its max function. The mix has a trigger and a graph curve. When the gpu temp rises, the gpu graph sets the gpu fans as needed. When the gpu temp drops, even if the gpu graph would be bellow its 0% fan speed point, but still higher than 42c, the trigger curve is active, holding the fans at 49%. Once the gpu temp drops below 42c, the idle point in the trigger curve is activated and the gpu fans turn off.

Easy!

**Comment by ionGL** [comment]
@Rem0o , i was testing a graph that i was thinking should work like a trigger.

![cooldown graph test](https://github.com/Rem0o/FanControl.Releases/assets/80151979/99af9cb0-8601-4797-bfe7-8b777fc9bde0)

First point at 52C, 0%
Second point at 53C, 49%

With a 10C hysteresis, the 0% should be activated once the temp drops to 43C, right?

Right now it doesn't seem to work that way. It goes to 0% as soon as it dropped 10C from where it was when the gpu load stopped.
Even if the graph is flat, with the only point at 53C, it saw 60C for example, and once it dropped 10C from there, it stopped the fan.
In a graph like this, flat, shouldn't that only point be latched and have the hysteresis calculated from its value before setting the previous % value in the graph, 0% in this case?

**Reply by Rem0o** [reply]
The graph hysteresis doesn't care about the shape or the number of points. It simply do its thing from the last state it was updated to. Most graph aren't flat, so they update from somewhere on a ramp, between points.

**Reply by ionGL** [reply]
Yes, it's a different hysteresis logic compared to what argus monitor uses for example.

Check this out. In argus, a similar graph, with a point at 53C and a hysteresis of 10.

![argus test](https://github.com/Rem0o/FanControl.Releases/assets/80151979/e650303e-4ab0-4d8e-a22e-17eb0836f85c)

Even if the temp went over 60C, it held that 50% and only applied 0% at exactly 42C.

It might make sense to implement a logic for sudden temp drops.

If i have specifically put a point on the graph (49% at 53C in my example from earlier) and the temp drops suddenly from way higher, i think hysteresis should take note of that user added point and adjust itself according to it.

**Reply by ionGL** [reply]
Did another argus test with a non flat curve.

![argus test2](https://github.com/Rem0o/FanControl.Releases/assets/80151979/e23e72e6-724a-43bb-8ccc-becad0b5b7c5)

It followed the graph and set the fan at 55% at 60C, then as the load was killed and temp dropped more than 10C instantly, it first reverted to 50%, then waited for 42C to set the fan at 0%.

**Reply by Rem0o** [reply]
Ahh I see, there's a small hysteresis "box" around each points, and the points are "sticky". Yeah it is a completely different approach. Intuitively from 55 with a 10 degree drop (exceeds hysteresis), I would of expected immediately the point at 45 deg, so 0%. I don't know why you would want to have it stick to the 52 degree point +- a 10 degree hysteresis ? I think I would receive far more confused user messages with that logic than the one currently in FanControl. 

To be clear, here's the FanControl logic: 

Instead of having a hysteresis "box" drawn around each point on the curve, Fancontrol has the box drawn around the current output of the graph:
![image](https://github.com/Rem0o/FanControl.Releases/assets/14118956/ba753bff-41b1-400d-9e73-0e525212168a)
In this image the last change to the output was at 65 degree. For a 5 degree hysteresis, it will take either 60-- or 70++ to trigger a change.  If the temp drops suddenly to 42, the box will be drawn around the 42 point, and it will take either 47 or 37 to trigger the next change and so on.... There is no stickiness, and the points are just a way to express a curve/function, there is no trigger or limit associated with them.

**Reply by ionGL** [reply]
Yes, argus hysteresis works differently.

"Hysteresis only applies only on the way down" is the default behavior and there is no option to have hysteresis on the way up . On temp rise, the % increases immediately, following the graph.

There are 7 hard coded graph points. You can play with them as you like, but you can't remove or add more.

I simulated with a graph like this and for temp source a 20 seconds gpu average, for it to see a slower cooldown
![argus test3](https://github.com/Rem0o/FanControl.Releases/assets/80151979/ca6fa9b4-8652-49d6-954d-c7fd63463db5)

52C - 0%
53C - 49%
54C - 50%
55C - 51%
56C - 52%
57C - 53%
58C  -54%

As the gpu load was killed from 65C, it went through all the % on the way down... 53% at 47C, 52% at 46C...... 49% at 43C and turned off the fan at 42C.

I saw the exact behavior with a curve like this, without custom points over 53C (ignoring that 120C).

![argus test4](https://github.com/Rem0o/FanControl.Releases/assets/80151979/13e2e613-2873-4573-9e48-ebc3b8fd53d6)

As temp goes down, **it sets new fan speeds relative to the hysteresis value and the temp it started going down from**. 1% at a time, if the cooldown is slow enough, simulated with an average over time.

If i simulate a jump from 40C to 55C then kill the load, the temp instantly drops more than 10C hysteresis, but it will still set the % to that 49% at 53C point it saw on the graph and waits for 42C to turn off the fan.

**Reply by ionGL** [reply]
Maybe i now understand argus hysteresis logic.

As temp goes down, **it follows all the graph's percentages in reverse, even the points between user selections, subtracting the hysteresis temp value.**

If on the way up a graph looks like this:

52C - 0%
53C - 49%
54C - 50%
55C - 51%
56C - 52%
57C - 53%
58C -54%

With a 10C hysteresis, on the way down it will set fan % like this:

47C - 53%
46C - 52%
45C - 51%
44C - 50%
43C - 49%
42C - 0%

At each sample interval it checks the fan % and temp and will not allow a lower fan % value if the temp doesn't drop at least the hysteresis value, starting the "follow graph in reverse mode". This is why a sudden temp drop from 58C to 45C will not trigger 0% in my example.

@Rem0o , what do you think?
